auth-api/src/main/java/io/stargate/auth/api/resources/AuthResource.java

```diff
@@ -232,16 +252,53 @@ public Response createTokenFromUsername(
     } catch (UnauthorizedException e) {
       return Response.status(Response.Status.UNAUTHORIZED)
           .entity(new Error("Failed to create token: " + e.getMessage()))
+          .cacheControl(cacheControlNoStore)
           .build();
     } catch (Exception e) {
       logger.error("Failed to create token", e);
       return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
           .entity(new Error("Failed to create token: " + e.getMessage()))
+          .cacheControl(cacheControlNoStore)
           .build();
     }
 
     return Response.status(Response.Status.CREATED)
         .entity(new AuthTokenResponse().authToken(token))
+        .cacheControl(cacheControlWithMaxAge)
         .build();
   }
+
+  /**
+   * Used to create a Cache-Control header containing the proper values for the token's max-age and
+   * s-maxage as defined by {@code stargate.auth_tokenttl} minus some time to avoid expiration. If
+   * the property is not set then the default value of 1790 will be used.
+   *
+   * @return A {@link CacheControl} object which is converted into a Cache-Control header with
+   *     proper directives.
+   */
+  private CacheControl cacheControlWithAge() {
+    // For a large enough TTL it's safe to knock off 10 seconds but for a smaller number reducing by
+    // that much will negatively impact caching so instead subtract only 1 second.
+    int maxAge = tokenMaxAge <= 100 ? tokenMaxAge - 1 : tokenMaxAge - 10;
+    CacheControl cacheControl = new CacheControl();
+    cacheControl.setMaxAge(maxAge);
+    cacheControl.setSMaxAge(maxAge);
```

Comment:
In general, caching may be tricky in this context. If the client gets a value from the cache, Stargate will _not_ update the row, so the data will not be refreshed and the TTL will apply to the original insert (assuming no on-going validation calls). Consequently, if the client attempts to validates the cached value after a period close to the TTL, the token row may already be expired, which will violate the TTL contract from the client's perspective.

I understand we need to _disable_ the caching error response. However, It is really necessary to _enable_ the caching of positive responses?

Reply:
Since the token is on a sliding window, in all actually it will be valid _longer_ than the max age. And I believe this will be the flow on the client side


10:00: createToken call, max age = 1790
10:20: Make request requiring token, check cache if token is still valid or a new one is needed. Still valid so use cached value
10:30: token row expires (according to client's caching logic)
10:40: Client makes another request and checking the cache sees the token is "expired" and calls createToken again

Although the logic will still hold with your scenario since at 10:20 when they call createToken again they will get a fresh token from stargate with a new TTL and corresponding max-age header.

Reply:
> Although the logic will still hold with your scenario since at 10:20 when they call createToken again they will get a fresh token from stargate with a new TTL and corresponding max-age header.

Not sure I understand. Why does it have to be a new token if cache settings allow reusing the old value?

Reply:
Also, my example is not about how the flow proceeds normally, but how it can go wrong and produce unexpected results.

Reply:
> Not sure I understand. Why does it have to be a new token if cache settings allow reusing the old value?

Do you mean "call" createToken but actually get the value out of the cache? If that's the case it's going to depend on how the client implements caching logic, either setting TTL at the cache level or thing-calling-cache-level. If it's at the cache level then should still be okay since I'd expect the cache to respect the max-age value. If it's at the thing-calling-cache-level then I'd hope it's recomputing the TTL otherwise that sounds like a bug waiting happen since there's a cache TTL disconnect between cache and the caller of the cache.

> Also, my example is not about how the flow proceeds normally, but how it can go wrong and produce unexpected results.

Sure but I think the best we can do is follow conventions of the Cache-Control header and hope they implement caching properly on their side. And honestly these tokens will last longer than they think due to the sliding window.

Reply:
I do believe even a properly configured cache can lead to unexpected behaviour here, because the HTTP call to create a token may return previous values within the "maxage" timeframe without hitting the backend, hence without updating the row timestamp in storage. Validation calls come though a different API and the HTTP cache is not aware of that.

What do you think about using "no-store" for _both_ successes and failures in token-related Web APIs?

Reply:
> because the HTTP call to create a token may return previous values within the "maxage" timeframe without hitting the backend

Sure but I don't see how that's a problem. We return the max-age in order to tell the cache to evict a token once it hits that age. If there's multiple tokens within 0 and 30 minutes old that's fine because they're still valid. It's up to the cache implementation to evict tokens based on `max-age` directive before they expire. If someone wants to set up a mutli-layer cache then I'd say the onus is on them to ensure that each of their layers is following caching standards as we are here by returning the Cache-Control header. 


> Validation calls come though a different API and the HTTP cache is not aware of that.

Sure and they shouldn't care. The job of the cache is to continue serving up the token until it needs to be evicted based on the prescribed TTL.


> What do you think about using "no-store" for both successes and failures in token-related Web APIs?

Then we'd be telling users to never cache a token. By default tokens have a TTL of 30 minutes which should be a perfectly acceptable time period to cache.

Reply:
Well, caching is different from TTL. We're telling caches not to cache, but end user know the token is valid for as long as the TTL. However, authentication-related interactions should normally not be cached. My reading of [RFC-7234](https://tools.ietf.org/html/rfc7234#section-3.2) is that it is the default standard behaviour, however, we explicitly allow caching in this case by providing cache-control headers.

Reply:
> A shared cache MUST NOT use a cached response to a request with an Authorization header field

That's referring to caching the response of an authenticated request. So for example, if someone were to call Stargate and get a row that technically shouldn't be cached because it's an authenticated request (granted we don't use the `Authorization` header but same concept). 

I'm pretty sure that section of the RFC is there to keep people honest. If I revoke your access to a resource you shouldn't have access to it anymore but you will if you've cached it. This doesn't really apply to our tokens since they're revalidated on every request which means if I revoke your token even if you have it cached it still won't do you any good since it's being revalidated.